### ISSUE TITLE
AttributeError from ObservationTable mask subset selection

### ISSUE THREAD

**Issue by mapazarr** [gammapy/gammapy#293]
@cdeil @adonath 
I have found a strange behavior in `~gammapy.obs.ObservationTable`. Im principle is `~ObservationTable` an `~astropy.table.Table`, so I would expect the same behavior, when using `~Table` methods, right?

In the following example script I try to apply a mask to an observation table (in this case a box in zenith angle):
 https://github.com/mapazarr/astropy_scripts/blob/master/astropy_scripts/test_observation_table.py.
I implemented 2 methods, the difference being the function used for creating the observation table:

 - In method 1 (works fine) `obs_table1` is actually a `~astropy.table.Table`.
 - In method 2 (fails!) `obs_table2` is a `~gammapy.obs.ObservationTable` (*).

(*) Method 2 works if I convert the `~ObservationTable` into `~Table` with `Table(np.array(obs_table))`, so I assume that not all the `~Table` functionality is available for `~ObservationTable`.

Could you tell me why does method 2 fail? Is there something missing in `~ObservationTable` (for instance a `__init__` and/or `copy` method)? (BTW, I tried to add them but it still didn't work, but maybe I did smthg wrong.)

Otherwise, a workaround could be to transform `~ObservationTable`s into `~Table`s for such problematic operations such as applying a mask, and afterwards convert back into `~ObservationTable`s. This option is not very elegant, but it would be fast to implement for now. What do you think?

**Comment by cdeil** [comment]
I'm not sure is this is the issue you are having, but here's one bug:
```python
>>> from gammapy.datasets.make import make_test_observation_table
>>> table = make_test_observation_table(observatory_name='HESS', n_obs=3)
>>> mask = table['OBS_ID'] == 1
>>> table[mask]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/deil/Library/Python/3.4/lib/python/site-packages/astropy-1.1.dev12736-py3.4-macosx-10.10-x86_64.egg/astropy/table/table.py", line 885, in __getitem__
    return self._new_from_slice(item)
  File "/Users/deil/Library/Python/3.4/lib/python/site-packages/astropy-1.1.dev12736-py3.4-macosx-10.10-x86_64.egg/astropy/table/table.py", line 551, in _new_from_slice
    table.meta.clear()
AttributeError: 'NoneType' object has no attribute 'clear'
```

@mapazarr – I'll leave the further debugging and fixing and adding a test case to you.
I have to go now and I don't really understand yet what's wrong.
Somehow the table we construct is corrupt ... either because of a bug in `ObservationTable` or in `make_test_observation_table` ... there can also be multiple issues, I don't have much experience yet with sub-classing `Table`.

The traceback mentions that `table[mask]` calls `_new_from_slice`, which accesses `table.meta` and that is `None` at that point for some reason.
One point is that `table.meta` should be an `OrderedDict` and not a `dict`:
https://github.com/gammapy/gammapy/blob/master/gammapy/datasets/make.py#L106
That's probably not the solution though ... my guess would be that we need to do something in the `ObservationTable` class definition to make it behave correctly.
So you could try to narrow this bug down to an even simpler test-case which doesn't involve `make_test_observation_table`, just `ObservationTable`.
And then play around with this and read code, or consult the Astropy docs on Table sub-classing.

@mapazarr – Please don't spend too much time on this. If you can't figure it out within an hour or two, make a comment here and I'll have a look on Monday, and if I also can't figure it out within an hour or two, we'll ask on astropy-dev for help.

**Comment by cdeil** [comment]
Just one more note ... `astropy.table.Table` now has a useful `info` method ... we have to rename the `info` method in `ObservationTable` and any other `astropy.table.Table` sub-classes to avoid overwriting it.

**Comment by mapazarr** [comment]
OK, I found the problem. It's a bug in the header definition in `make_test_observation_table`. It doesn't matter if it's defined as `dict` or `OrderedDict`.
But this line:
 https://github.com/gammapy/gammapy/blob/master/gammapy/datasets/make.py#L106
Should be
 `obs_table.meta = header`
I'll commit a fix soon.

**Comment by cdeil** [comment]
@mapazarr – Thanks for figuring that out!

In this case, you can just put a commit with this fix in the next PR you're working on, no need for a separate one. The Astropy policy is to leave issues open until the fix has landed in master ... let's do the same.

**Comment by cdeil** [comment]
Oh, ... and even if `dict` vs. `OrderedDict` wasn't the issue ... please change to an `OrderedDict` ... you can use this pattern to fill it in an order that's then reflected in the FITS header it generates:
```
meta = OrderedDict()
meta['this'] = 42
meta['that'] = 43
```
I also like it because it's very readable (even if longish sometimes) and it's easy to add / remove / comment lines with simple diffs in version control.

**Comment by mapazarr** [comment]
Well, I did smthg different: I directly define it in the table, without the need to specify what kind of dictionary it is:
`obs_table.meta['ONE'] = 1`
This is less error prone in case the type of dict changes in `Table`.
I also added a header test in `test_make_test_observation_table`.

**Comment by mapazarr** [comment]
The fix is included in PR #295.
